### PR TITLE
Sequence/regex schema

### PR DESCRIPTION
**Pull request by green-coder**
This is a proposal for the format of data the users will provide to Malli. (Related to #180)

Notes:
- I think that users of `:cat` will use it more often *with* names on children rather that *without*.
- I chose to make the format anonymous/non-anonymous explicit, to avoid conflicts between names of children and other operators like `:*` or `:alt`.

### CODE REVIEW

File: test/malli/core_test.cljc [hunk 1]
```diff
@@ -586,6 +586,74 @@
                                         [false [:sequential int?] #{1 2 3}]
                                         [false [:sequential int?] nil]]
 
+                          "cat" [; same as [:cat {:into nil} ...]
+                                 [true [:cat [:x int?]] '(1)]
+                                 [true [:cat [:x int?]] [1]]
+
+                                 ; aka :cat*
+                                 [true [:cat {:into :any} [:x int?]] '(1)]
```
Comment (ikitommi) [inline test/malli/core_test.cljc]: does `:into` mean just type validation? If so, the name is bit misleading.

File: test/malli/core_test.cljc [hunk 2]
```diff
@@ -586,6 +586,74 @@
...
+                                 ; concatenated sequences
+                                 [true [:cat [:a [:cat [:ax int?] [:ay int?]]]
+                                             [:b [:cat [:bx int?] [:by int?]]]] [1 2 3 4]]
+
+                                 ; sequence hierarchy
+                                 [true [:cat [:a [:cat {:into :any} [:ax int?] [:ay int?]]]
+                                             [:b [:cat {:into :any} [:bx int?] [:by int?]]]] ['(1 2) [3 4]]]
+
+                                 ; aka :acat
+                                 [true [:cat {:anonymous true} int? int? [:* string?]] [1 2 "a" "b"]]
```
Comment (ikitommi) [inline test/malli/core_test.cljc]: currently, both `:or` and `:and` are anonymous by default, should `:cat` be too? Once there is some real use for the named branches, I though of adding variants supporting with names. My initial hunch was to use different schema names, something like:

```clj
;; anonymous
[:or int? string?]

;; named
[:orn [:i int?] [:s str?]]

;; .. or via props
[:or {:named true} [:i int?] [:s str?]]
```

If there are two different things that contribute to schema names, it starts to be awkward:

```clj
;; anonymous
[:cat ...]

;; named branches
[:catn ...]

;; named + vector
[:catnv ...]

;; anonymous + vector
[:catv ...]
```

my 2 cents this morning: the "named branches" would go to name (`:cat` & `:catn`) and the type could be a property. What do you think?

**Comment by ikitommi** [comment]
> I think that users of :cat will use it more often with names on children rather that without.

I think so too, but would be good to be coherent with `:or`. Well, pre-alpha, so we can change things.

File: test/malli/core_test.cljc (hunk 1 above)
Comment (green-coder) [reply test/malli/core_test.cljc]: It is meant for validation and generation. I took this wording from an option of https://clojure.org/guides/spec#_collections and would not mind changing it.

File: test/malli/core_test.cljc (hunk 2 above)
Comment (green-coder) [reply test/malli/core_test.cljc]: The names on `:and`, `:or`, `:cat` and `:alt` do not provide the same kind of information to the user:
- Names on `:and` and `:or` help reporting on which predicates invalidate (`:and`) or validate (`:or`) the data. They are used for indicating **properties** of the data. I guess that the user does not need this information most of the time.
- Names on `:cat` and `:alt` are providing the information on the **identity** of the data. Users would need it most of the time.

To me, it does not look strange for Malli to have by default no names on `:and` and `:or` and have some on `:cat` and `:alt`. Clojure Spec also have a mix of the 2, `s/and` has no name while `s/or`, `s/cat` and `s/alt` have them.

`{:anonymous true/false}` and `{:named: false/true}` encode the same information, I prefer `{:named false/true}` as it is shorter to read and write.

**Comment by green-coder** [comment]
> Tests look great! As the destructuring is a valid application for sequence/regex schemas, ideas how to implement the s/conform with malli?
>
> My initial thought was to make m/explain explain also successful values, so one would get both the values and the branching information from it's result (if not errors).

I see 2 different usages there:
1. The human who occasionally wants to get information of what went wrong in a validation, possibly via a REPL.
2. The program who would do serial data transformation for further processing and that would not be interested to know why some of the transformation failed.

In the future, the API for `1.` would likely evolve toward better explanations, even if it costs more CPU, while the API for `2.` would likely evolve toward something that run faster. For this reason, I would recommend having different functions for the 2 usages. In other words, I think that explain, conform and unform should be separated in the API. They could however use common functions in their implementation.

At the moment, I don't know where would be the best place for the implementation of conform. Maybe as new functions in the Schema protocol, or as a Transformer, or as something else.

**Comment by green-coder** [comment]
I am going to implement a naive implementation first: Because its implementation will be simple it will be easier to adapt to support recursive models later on.

File: test/malli/core_test.cljc (hunk 1 above)
Comment (green-coder) [reply test/malli/core_test.cljc]: @ikitommi Do you have a suggestion for an alternative to `:into`?

File: test/malli/core_test.cljc [hunk 3]
```diff
@@ -586,32 +586,32 @@
...
-                                 [true [:cat {:into :list} [:x int?]] '(1)]
-                                 [false [:cat {:into :list} [:x int?]] [1]]
+                                 [true [:list-of [:cat [:x int?]]] '(1)]
+                                 [false [:list-of [:cat [:x int?]]] [1]]
 
-                                 ; aka :catv
-                                 [false [:cat {:into :vector} [:x int?]] '(1)]
-                                 [true [:cat {:into :vector} [:x int?]] [1]]
+                                 [false [:vector-of [:cat [:x int?]]] '(1)]
+                                 [true [:vector-of [:cat [:x int?]]] [1]]
```
Comment (ikitommi) [inline test/malli/core_test.cljc]:
```clj
[:vec-of [:cat [:x int?]]
```
can be done with:
```clj
[:and vector? [:cat [:x int?]]
```

Comment (green-coder) [reply test/malli/core_test.cljc]: That's fixed in the new commits.

**Comment by green-coder** [comment]
Things done in this PR:
- Some tests for the validators of the sequence schemas.
- Validators for the sequence schemas are implemented and working.

Things to be done in this PR:
- Finish the implementation for the transformers,
- review the LensSchema implementation which may not be appropriate (copied from -map-schema),
- review the implementation of the visitors (copied from -map-schema),
- add generators for the sequence schemas.

**Comment by ikitommi** [comment]
thanks once more, the clean slate impl is around the corner.

**Comment by green-coder** [comment]
You're welcome, I am happy it helped.

**Pull request closed without merging**